### PR TITLE
Step 7 to step 8

### PR DESCRIPTION
**Pull request by giuscri**

### CODE REVIEW

File: src/main/scala/rps/GameRepository.scala [hunk 1]
```diff
@@ -5,19 +5,21 @@ import scala.collection.concurrent.TrieMap
...
-  private val map = TrieMap.empty[String, Play]
+  private val map = TrieMap.empty[Int, Play]
+  private var nextUserId = 0
 
-  override def save(play: Play): Unit = {
-    map.put("play", play)
-    ()
+  override def save(play: Play): Int = {
+    map.put(nextUserId, play)
+    nextUserId = nextUserId + 1
```
Comment (lvolp) [inline src/main/scala/rps/GameRepository.scala]: Probably will be better to have a specific case class for  game id, take  also a look to AtomicInteger for id generation

**Review by lvolp — CHANGES_REQUESTED** [review]

File: src/main/scala/rps/GameRepository.scala (hunk 1 above)
Comment (giuscri) [reply src/main/scala/rps/GameRepository.scala]: ~`case UserId(userId: Int)` seems overkill. Is that what you mean?~ Yes

File: src/main/scala/rps/GameRepository.scala [hunk 2]
```diff
@@ -3,21 +3,24 @@ package rps
...
 trait GameRepository {
-  def save(play: Play): Unit
-  def read(): Option[Play]
+  def save(play: Play): Int
+  def read(userId: Int): Option[Play]
 }
 
 class InMemoryGameRepository extends GameRepository {
-  private val map = TrieMap.empty[String, Play]
+  private val map = TrieMap.empty[Int, Play]
```
Comment (calippo) [inline src/main/scala/rps/GameRepository.scala]: Offtopic (not realted to this PR): beware `TrieMap` `getOrElseUpdate` operation is not thread safe. In Banksealer we prefer `ConcurrentHashMap`. Ask Ema for details.

**Review by calippo — APPROVED** [review]

**Comment by giuscri** [comment]
On Mon, Nov 11, 2019 at 11:25 AM Claudio Caletti <notifications@github.com>
wrote:

> Offtopic (not realted to this PR): beware TrieMap getOrElseUpdate
> operation is not thread safe. In Banksealer we prefer ConcurrentHashMap.
> Ask Ema for details.
>
As I understand this, `TrieMap.getOrElseUpdate` should be thread-safe now
(as of scala 2.11.6 [1]) /cc @cardu

---
[1] https://github.com/scala/scala/pull/4319

File: src/main/scala/rps/GameRepository.scala (hunk 2 above)
Comment (giuscri) [reply src/main/scala/rps/GameRepository.scala]: As I understand this, `TrieMap.getOrElseUpdate` should be thread-safe now
([as of scala 2.11.6](https://github.com/scala/scala/pull/4319)) /cc @Cardu

Comment (calippo) [reply src/main/scala/rps/GameRepository.scala]: As far as I remember `TrieMap` doesn't have a `computeIfAbsent` equivalent. That makes the semantic a little odd. But leave it to @Cardu

Comment (Cardu) [reply src/main/scala/rps/GameRepository.scala]: TrieMap is thread safe regarding it's own structure, the problem with this is that the expression that computes the value may be called multiple times if there are more than one actor that concurrently tries to insert an element that was not present.

So you have to be really careful to not cause any side effect in the expression that you pass to this method, for example, we were creating a table on the database in the evaluation, and sometimes we had exception for `table already existing error`

This is also stated in a comment in scala source code: https://github.com/scala/scala/blob/7986814ead106fc7edf82eb792878ca6f056df44/src/library/scala/collection/concurrent/TrieMap.scala

(just before `getOrElseUpdate` method)

**Pull request merged**